### ISSUE TITLE
Display out of stock products at the end of the catalog.

### ISSUE THREAD

**Issue by yofisim** [magento/magento2#7589]
I would like to display the out of stock products at the end of my catalog.
This functionality was available in M 1.9 (and maybe before), and for some reason it is not available anymore, not in system configuration nor store configuration options.

### Preconditions

1. Magento 2.1.1
2. Apache Webserver
3. PHP 7.0

### Steps to reproduce

1. Have a catalog with products, some in stock, and others out of stock.
2. Try to display the out of stock products at the end of the catalog, on category products, or homepage.

### Expected result

1.  A functionnality to allow out of stock products to be displayed at the end.

### Actual result

1. In stock and out of stock products are melted together.

**Comment by odubovyk** [comment]
Cannot reproduce the issue, "out of stock" products are at the end.
Write more detailed steps to reproduce.

**Comment by yofisim** [comment]
@odubovyk It's actually the issue. Out of stocks products and in-stock products are on mixed together. I want to display out of stock products at the end of the catalog.

Please check here: http://vps307618.ovh.net/chaise-chesterfield

You will see the out of stock products mixed with in stock products.

**Comment by odubovyk** [comment]
I believe, current info is not enough, as I've said we need more details of steps to reproduce.
Please see steps to reproduce section in https://github.com/magento/magento2/wiki/Issue-reporting-guidelines.

**Comment by veloraven** [comment]
According to [contributor guide](http://devdocs.magento.com/guides/v2.1/contributor-guide/contributing.html), tickets without response for two weeks should be closed.
If this issue still reproducible please feel free to create the new one: format new issue according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines): with steps to reproduce, actual result and expected result and specify Magento version.

**Comment by MMaKenZi** [comment]
i think this issue is so important,
the meaning of this issue is this : 
Show in stock products first and out of stock products last

i will explain by an example


it happened sometimes that whole products of a category are out of stock, by the only option provided by magento, they all will be hidden, so if a user try to browse site by categories menu, he will just see "no products found!" and say what a stupid shop is this that created some empty categories!

but if the user be able to see the out of stock products, can contact admin for some other products and don't think about the stupidity of the shop anymore :D

this example was just for importance of out of stock products,
but think you are in a category which on it's first page, you have for example 10 in stock products and 10 out of stock products, absolutely a customer which enter a shop, just want to see the in stock products, so they should be listed first, and as I explained the importance of out of stock products, they should be listed at last

so I think the best option is this:
first in stock products be visible and after them out of stock products

is this clear now?



My tested version is 2.1.9
PHP is 7.0.22
with nginx webserver

**Comment by MMaKenZi** [comment]
Shouldn't be opened again?

**Comment by orlangur** [comment]
@MMaKenZi, nope, as far as I can say.

GitHub is intended for bug reports, please use [Community Forums](https://community.magento.com/t5/Magento-2-Feature-Requests-and/idb-p/feature-requests) for other suggestions like improvements or feature requests.

**Comment by yofisim** [comment]
I left M2 for Prestashop, I think everyone knows the reason why.
It is definitely a bug, not a feature request. I mean who will want to display out of stock products, who cannot be ordered, mixed with in-stock products.

Anyway, a nasty trick I did in the past was to add a "custom field" (don't remember the exact nomination for M2), to each product.

Each product who is in stock gets value 1
Each product who is out of stock gets value 2

Then you can sort the catalog (from the admin) to display products with custom field value == 1 first, and all the others after.

If you have a lot of products, like us in the past (more than 50k), you should then make a CSV import with the field already set and value applied.

Matters of hours, but was working pretty well back in time.

Good luck

**Comment by orlangur** [comment]
> Each product who is in stock gets value 1
> Each product who is out of stock gets value 2

That sounds like a hack, it is quite easy to achieve desired behavior with custom implementation (since stock information is already in collection, you just need to add additional sorting).

IT IS a feature request in fact, although such feature is quite easy to implement, not sure it would be in-demand in core. Quite often out-of-stock products are simply not displayed (there is an option for this) or backorders are enabled so that stock status does not really matter.

**Comment by yofisim** [comment]
I understand your saying, though I don't agree at all with you.
If a company using SEO as a traffic acquisition way, they HAVE to display out of stock products, doesn't matter if backorders are enabled or not.

Depending on how you are using Magento, it is a BUG or a FEATURE REQUEST.
As far as the impact of the problem is big (and old) enough, this is something that should be corrected.

I will do the code changes and post them here tonight, it might be useful for some people who can't wait for it.

**Comment by orlangur** [comment]
> I will do the code changes and post them here tonight, it might be useful for some people who can't wait for it.

Oh, just post a PR so that it can get merged into core (targeting `2.2-develop` if it is not breaking BC). We need a PO approval to add features but as you can provide a real-world use case it should not be a problem.

**Comment by mariommoreno** [comment]
We have the same needs. Any progress with this feature?

**Comment by oviliz** [comment]
Have you had any chance to do that @yofisim ?

**Comment by Sanfam** [comment]
Not to create an "Angry Mob" Echo Chamber, but I'd love to see what you've come up with, @yofisim !  I've been hoping to do this exact modification on my instance and am eager to see how it can be achieved.

**Comment by yofisim** [comment]
Hi guys,

unfortunately no, it was more complicated than I expected, and I left Magento to be fairly honest.
But again, I do remember how I tricked this to make it work. Just check my previous comments.

If it is not clear, here is how I achieve it (don't pay attention to the terms I use, as they are probably not the exact name on Magento. Just get the process, than replicate) :

1) I created a "custom attribute" for all my products called stock.
2) Using CSV import, I added to the stock field:
- value "1" if the product is in stock.
- value "0" if the product is Out Of Stock.

3) Once import is finished, when you are on the catalog product view in the admin, you can customize the sorting of the catalog. Apply the sorting on the custom attribute "stock", with the value ASC (ascending).

4) Done. All your products will be sorted with in-Stock first, and then out of stock, at the end of the catalog.

Hope this helps.

Edit: as @orlangur stated, it is definitely a hack, and I suggest you to find better solutions to manage your products displaying. It was working for me, because I had endless stock on "in-stock products". And when a product is out-of-stock, it was for a long time. Adapt to your situation.

**Comment by orlangur** [comment]
@yofisim thanks for described approach. Yeah, something like that should be possible with a custom implementation on top of collection query instead of an additional attribute like you achieved it.

**Comment by JuanjoR** [comment]
however, it is not the solution in configurable products, since the configurable products do not carry independent stock.
In my store, if I used that method, all configurables would always be at the end, because when importing the stock, the configurables do not change with stock / without stock

in M1 I used this module that worked to perfection, but I do not find something similar in M2
https://github.com/cgzhang/Rayfox_Catalog

**Comment by bengower** [comment]
I created something to do this:
https://github.com/bengower/OosBottom
It's loosely based on the Rayfox_Catalog module

**Comment by alvaialex** [comment]
I found a extension for the same

[Out Of Stock Last for M2](https://marketplace.magento.com/igorludgero-outofstocklast.html)

I hope this will be help.

**Comment by inkobject** [comment]
alvaialex, unfortunately the extension you cite was not compatible with my 2.2.3 install. It trashed my front end. Oh well, at least I only installed it on a dev environment 

Thanks for the suggestion, though.

**Comment by visahardik** [comment]
@bengower your suggested extension does not support while change drop down value in toolbar

**Comment by JKetelaar** [comment]
Simple fix:
In your `di.xml` add:
```
    <type name="Magento\Catalog\Model\Layer">
        <plugin name="orderOutOfStock" type="Vendor\Module\Plugin\Model\Layer" sortOrder="999"/>
    </type>
```

Then your `Layer.php`:
```
<?php
/**
 * @author JKetelaar
 */

namespace Vendor\Module\Plugin\Model;

use Magento\Catalog\Model\ResourceModel\Product\Collection;

class Layer
{
    /**
     * @param \Magento\Catalog\Model\Layer $subject
     * @param Collection $collection
     * @return mixed
     */
    public function afterGetProductCollection(\Magento\Catalog\Model\Layer $subject, $collection)
    {
        $collection->getSelect()->order('is_salable DESC');

        return $collection;
    }
}
```

This works both with changing the order in the frontend (as @visahardik mentioned) and with custom modules, such as the Amasty Improved Layered Navigation module.

**Comment by doublesharp** [comment]
@JKetelaar using `afterGetProductCollection()` applied the sort order several times for me, I had better luck with `beforePrepareProductCollection()` which is called once in `getProductCollection` and then cached - 

```
  /**
   * Sort items that are not salable last
   *
   * @return \Magento\Catalog\Model\ResourceModel\Product\Collection
   */
  public function beforePrepareProductCollection(
    \Magento\Catalog\Model\Layer $layer,
    \Magento\Catalog\Model\ResourceModel\Product\Collection $collection
  ) {
    // apply sort order
    $collection->getSelect()->order('is_salable DESC');

    // make sure this sort is first applied to the results
    $orders = $collection->getSelect()->getPart(\Zend_Db_Select::ORDER);
    array_unshift($orders, array_pop($orders));
    $collection->getSelect()->setPart(\Zend_Db_Select::ORDER, $orders);

    return [$collection];
  }
```

**Comment by faisal-project** [comment]
> @JKetelaar using `afterGetProductCollection()` applied the sort order several times for me, I had better luck with `beforePrepareProductCollection()` which is called once in `getProductCollection` and then cached -
> 
> ```
>   /**
>    * Sort items that are not salable last
>    *
>    * @return \Magento\Catalog\Model\ResourceModel\Product\Collection
>    */
>   public function beforePrepareProductCollection(
>     \Magento\Catalog\Model\Layer $layer,
>     \Magento\Catalog\Model\ResourceModel\Product\Collection $collection
>   ) {
>     // apply sort order
>     $collection->getSelect()->order('is_salable DESC');
> 
>     // make sure this sort is first applied to the results
>     $orderBys = $collection->getSelect()->getPart(\Zend_Db_Select::ORDER);
>     array_unshift($orderBys, array_pop($orderBys));
>     $collection->getSelect()->setPart(\Zend_Db_Select::ORDER, $orderBys);
> 
>     return [$collection];
>   }
> ```

Thanks @doublesharp, working fine even with price sorting filter as well.

**Comment by umar-emo** [comment]
@doublesharp OR @faisal-project 
Can you explain more clearly How you guys implement this??


Thanks,

**Comment by doublesharp** [comment]
@umar-emo You need to create a module (or add to another module) the plugin that modifies the behavior - 

**`magento/app/code/Doublesharp/SortOutOfStockLast/registration.php`**
```
<?php
\Magento\Framework\Component\ComponentRegistrar::register(
    \Magento\Framework\Component\ComponentRegistrar::MODULE,
    'Doublesharp_SortOutOfStockLast',
    __DIR__
);
```

**`magento/app/code/Doublesharp/SortOutOfStockLast/etc/module.xml`**
```<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../../../lib/internal/Magento/Framework/Module/etc/module.xsd">
    <module name="Doublesharp_SortOutOfStockLast" setup_version="1.0.0" active="true">
        <sequence>
            <module name="Magento_Catalog"/>
            <module name="Magento_CatalogSearch"/>
        </sequence>
    </module>
</config>
```

**`magento/app/code/Doublesharp/SortOutOfStockLast/etc/di.xml`**
```
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../../../../lib/internal/Magento/Framework/ObjectManager/etc/config.xsd">
    <!-- sort out of stock items last -->
    <type name="Magento\Catalog\Model\Layer">
        <plugin name="catalog_model_category_sort_out_of_stock_last" type="Doublesharp\SortOutOfStockLast\Plugin\Catalog\Model\Layer" sortOrder="1000"/>
    </type>
</config>
```

**`magento/app/code/Doublesharp/SortOutOfStockLast/Plugin/Catalog/Model/Layer.php`**
```
<?php

namespace Doublesharp\SortOutOfStockLast\Plugin\Catalog\Model;

/**
 * Class Layer
 * @package Doublesharp\SortOutOfStockLast\Plugin\Catalog\Model\Layer
 */
class Layer
{
  /**
   * Sort items that are not salable last
   * prepareProductCollection
   *
   * @return \Magento\Catalog\Model\ResourceModel\Product\Collection
   */
  public function beforePrepareProductCollection(
    \Magento\Catalog\Model\Layer $layer,
    \Magento\Catalog\Model\ResourceModel\Product\Collection $collection
  ) {
    $collection->getSelect()->order('is_salable DESC');
    $orderBys = $collection->getSelect()->getPart(\Zend_Db_Select::ORDER);
    array_unshift($orderBys, array_pop($orderBys));
    $collection->getSelect()->setPart(\Zend_Db_Select::ORDER, $orderBys);

    return [$collection];
  }
}
```

**Comment by umar-emo** [comment]
@doublesharp Perfect. 

Let me make it as you explained.

May I know with that the filters are also working? Like sort by price, Position, Average etc

**Comment by doublesharp** [comment]
Yes, they all work for me. The module just uses a plugin to add an
additional sort to the query.

**Comment by umar-emo** [comment]
@doublesharp 
When I make an module as you explain, I got this error.

![12](https://user-images.githubusercontent.com/10205592/117190300-aebc4200-adf8-11eb-9b61-04b9c683c059.PNG)

**Comment by doublesharp** [comment]
@umar-emo that's from a different post, my example uses a different namespace/directory structure.

**Comment by umar-emo** [comment]
@doublesharp 
Brother, when I remove your module from the directory, composer run fine.
I thought, It happening due to the different directory structure is different.

**Is there any common way, which is suitable for all type of directory.?**
Looking forward to hearing from you.
Thanks

![Csapture](https://user-images.githubusercontent.com/10205592/117226129-63705680-ae2d-11eb-9afb-0c217e44c290.PNG)

**Comment by ia-gaurav** [comment]
@doublesharp  - seems you module does not works with magento 2.4 

Query i see using it in layer.php --> **SELECT `e`.* FROM `catalog_product_entity` AS `e` ORDER BY `is_salable` DESC**

But there in not column is_salable in catalog_product_entity database

**Comment by umar-emo** [comment]
Hy @doublesharp 

I found that the perfect solution, Its working on all magento2 versions.
Even Filters also working fine.

https://meetanshi.com/blog/sort-magento-2-category-products-by-stock-quantity/amp/

**Comment by ia-gaurav** [comment]
@umar-emo  this seems not working in magento 2.4

**Comment by web5exposure** [comment]
Also not working for me in magento 2.4.2 with elasticsearch

**Comment by Bashev** [comment]
Released in 2.4.5 with a lot of another bugs, related to it. Probably is a good idea to be used after 2.4.5-p1
